### PR TITLE
DOC: Add validation error when a docstring is missing 

### LINKED ISSUE #23648: DOC: Add validation error when a docstring is missing
We have the script `scripts/validate_docstrings.py` that checks several formatting errors in the docstrings.

Currently, when a function, method... has no docstring at all, it fails few of the checks, but we don't have one specific to it. See for example:
```
$ ./scripts/validate_docstrings.py --prefix=pandas.Series.imag
pandas.Series.imag: Docstring text (summary) should start in the line immediately after the opening quotes (not in the same line, or leaving a blank line in between)
pandas.Series.imag: Closing quotes should be placed in the line after the last text in the docstring (do not close the quotes in the same line as the text, or leave a blank line between the last text and the quotes)
pandas.Series.imag: No summary found (a short summary in a single line should be present at the beginning of the docstring)
```

What would be useful is to add a new check, that reports that the function does not have a docstring. And in the previous errors, check when the docstring is empty, and do not report them. So the previous call should report a single error with a text like `pandas.Series.imag: No docstring found`.

### PR DESCRIPTION
**Pull request by avolkov**
- [x] closes #23648
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/base.py [hunk 1]
```diff
@@ -668,7 +668,7 @@ class IndexOpsMixin(object):
     __array_priority__ = 1000
 
     def transpose(self, *args, **kwargs):
-        """ return the transpose, which is by definition self """
+        """ Return the transpose, which is by definition self """
```
Comment (WillAyd) [inline pandas/core/base.py]: This is orthogonal to the change right? Would be better to keep these in a separate PR focused on fixing up these docstrings, especially given there are other issues with the docstrings (ex: missing period)

File: scripts/tests/test_validate_docstrings.py [hunk 2]
```diff
@@ -668,6 +668,12 @@ def missing_whitespace_after_comma(self):
         pass
 
 
+class NoDocstrings(object):
```
Comment (WillAyd) [inline scripts/tests/test_validate_docstrings.py]: Does this need to be a dedicated class or can we just add this method to the existing class with bad docstrings?

File: scripts/tests/test_validate_docstrings.py [hunk 3]
```diff
@@ -720,6 +726,13 @@ def test_bad_class(self):
         assert isinstance(errors, list)
         assert errors
 
+    @capture_stderr
+    def test_bad_class(self):
```
Comment (WillAyd) [inline scripts/tests/test_validate_docstrings.py]: Hmm most of our focus in on docstrings for methods. Not sure we explicitly require the class to have a doctoring though. @datapythonista

File: scripts/tests/test_validate_docstrings.py [hunk 4]
```diff
@@ -720,6 +726,13 @@ def test_bad_class(self):
...
+        errors = validate_one(self._import_path(
+            klass='NoDocstrings'))['errors']
+        assert isinstance(errors, list)
```
Comment (WillAyd) [inline scripts/tests/test_validate_docstrings.py]: If we need this (though not sure we do) would be better to have a stronger assertion

File: scripts/validate_docstrings.py [hunk 5]
```diff
@@ -585,6 +586,20 @@ def validate_one(func_name):
 
     errs = []
     wrns = []
+
+    if len(doc.raw_doc) == 0:
```
Comment (WillAyd) [inline scripts/validate_docstrings.py]: Can we just leverage the implicit truthiness here instead, so `if not doc.raw_doc` or maybe even just `if not doc`?

File: scripts/validate_docstrings.py [hunk 6]
```diff
@@ -585,6 +586,20 @@ def validate_one(func_name):
...
+        errs.append(error('GL08'))
+        return {
```
Comment (WillAyd) [inline scripts/validate_docstrings.py]: Do we need to explicitly return here? Would be more inline with rest of the code if we allow it to pass over the remaining conditions (which would assumedly be all False anyway) and return at the end of the function

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/base.py (hunk 1 above)
Comment (avolkov) [reply pandas/core/base.py]: Yes, sorry about that, this got merged from the other branch I've been working on.

I added a revert commit, but I'm going to look for a way of removing the commit entirely

File: scripts/tests/test_validate_docstrings.py (hunk 2 above)
Comment (avolkov) [reply scripts/tests/test_validate_docstrings.py]: It's not a bad docstring It's missing docstring, out of these, each of the classes describe particular bad docstring behaviour: 

* BadGenericDocstrings -- badly phrased docstrings
* BadSummarries -- badly summarized docstrings
* BadParameters -- badly documented parameters
* BadReturns -- bad return docs
* BadSeeAlso -- badly formatted see also section
* BadExamples -- bad examples section

I guess this method might go to BadGenericDocstrings, but none of the candidates for it seem ideal.

File: scripts/tests/test_validate_docstrings.py (hunk 4 above)
Comment (avolkov) [reply scripts/tests/test_validate_docstrings.py]: on the second thought this test is not strong enough, adding a test case in test_bad_generic function should be sufficient.

File: scripts/validate_docstrings.py (hunk 6 above)
Comment (avolkov) [reply scripts/validate_docstrings.py]: The rest of the condition will not be False, if we don't do early return the next two conditions would evaluate to true

doc.start_blank_lines
doc.end_blank_lines

it's possible to put the code underneath into all-encompassing else statement, but the early return just seems cleaner.

File: scripts/tests/test_validate_docstrings.py (hunk 3 above)
Comment (avolkov) [reply scripts/tests/test_validate_docstrings.py]: I forgot to rename function i've been using as an example. will fix

File: scripts/tests/test_validate_docstrings.py (hunk 4 above)
Comment (avolkov) [reply scripts/tests/test_validate_docstrings.py]: an argument for leaving it in (with the fixed name) -- this class will give more informative error message in case NoDocstring class is removed from test source.

File: pandas/core/base.py (hunk 1 above)
Comment (avolkov) [reply pandas/core/base.py]: I've tried to do a rebase removing the commits from local tree that I then force-pushed to remote. That didn't seem to remove offending commits from PR.

@WillAyd do you mind if I create a new pull request once the rest of the issues have been rectified?

**Comment by WillAyd** [comment]
New PR is fine. I'll close this one out
